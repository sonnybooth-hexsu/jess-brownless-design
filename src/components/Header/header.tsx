import React from "react"
import { Link } from "gatsby"
import { Menu } from "react-feather"
import { CMSSubNav } from "./CMSSubNav"
import { SubMenuItem } from "./SubMenuItem"
import { useScroll } from "../../utils/functions"
import logo from "../../assets/icons/logo.svg"
import styles from "./header.module.css"

type HeaderProps = {
  siteTitle: string
  navLinks: Array<{
    name: string
    nameShort: string
    page: string
    id: number
    subLinks?: Array<{ name: string; page: string; id: number }>
  }>
  productItems: object
  navToggle: () => Function
  openMenu: number
  menuToggle: () => Function
  headerBg?: string
  headerColor?: string
  page: string
}

const Navigation = ({
  navLinks,
  openMenu,
  menuToggle,
  productItems,
  page: pageRendered,
}) => {
  return (
    <nav className={styles.headerNavigation}>
      <ol className={styles.headerNavigationList}>
        {navLinks.map(({ name, nameShort, page, subLinks, id, products }) => {
          if (products) {
            let data

            // @ts-ignore: graphql data
            if (products === "conditions") {
              data = productItems.conditions
            }

            // @ts-ignore: graphql data
            if (products === "oculoplasticServices") {
              data = productItems.oculoplasticServices
            }

            // @ts-ignore: graphql data
            if (products === "aestheticServices") {
              data = productItems.aestheticServices
            }

            data.edges.sort((a, b) =>
              a.node.childMarkdownRemark.frontmatter.id >
              b.node.childMarkdownRemark.frontmatter.id
                ? 1
                : -1
            )

            const productsDestructured = data.edges.map(
              obj =>
                // @ts-ignore: graphql data
                obj.node.childMarkdownRemark.frontmatter
            )

            return (
              <CMSSubNav
                page={pageRendered === products}
                key={`navigation-item-${id}`}
                name={name}
                nameShort={nameShort}
                title="Treatments"
                items={productsDestructured}
                id={id}
                open={openMenu === id}
                openTrigger={menuToggle}
              />
            )
          }

          if (subLinks && subLinks.length > 0) {
            return (
              <SubMenuItem
                key={`navigation-item-${id}`}
                name={name}
                subLinks={subLinks}
                id={id}
                open={openMenu === id}
                openTrigger={menuToggle}
              />
            )
          }

          return (
            <li
              className={`${styles.headerNavigationLinkItem} ${
                name === pageRendered ? `afterBorder` : ``
              }`}
              data-testid="headerNavigationLink"
              key={`navigation-item-${id}`}
            >
              <Link className={styles.headerNavigationLink} to={`${page}`}>
                {nameShort ? (
                  <>
                    <span className="xl:hidden">{nameShort}</span>
                    <span className="hidden xl:block">{name}</span>
                  </>
                ) : (
                  <span>{name}</span>
                )}
              </Link>
            </li>
          )
        })}
        <li
          className={`${styles.headerNavigationLinkItem} ${
            pageRendered === "Contact" ? `afterBorder` : ``
          }`}
        >
          <Link className={styles.headerNavigationLink} to="/contact">
            Contact
          </Link>
        </li>
      </ol>
    </nav>
  )
}

export const Header = ({
  siteTitle,
  navLinks,
  navToggle,
  openMenu,
  menuToggle,
  headerBg,
  headerColor,
  productItems,
  page,
}: HeaderProps) => {
  const { scrollY, scrollDirection } = useScroll()

  return (
    <div
      className={`
      ${headerBg ? headerBg : "bg-Brand-white"}
      ${headerColor ? headerColor : "text-Brand-black"}
      ${scrollY > 500 ? styles.headerContainerFixed : ""}
      ${
        scrollY > 500 && scrollDirection === "down"
          ? styles.headerContainerRetreat
          : ""
      }
       ${styles.headerContainer} fadedBorder`}
    >
      <div className={`${styles.header} `}>
        <Link className={styles.headerLogo} to="/">
          <img className={styles.headerLogoImage} src={logo} alt={siteTitle} />
        </Link>
        <Menu onClick={navToggle} className={styles.headerMenuIcon} />
        <Navigation
          productItems={productItems}
          navLinks={navLinks}
          openMenu={openMenu}
          menuToggle={menuToggle}
          page={page}
        />
      </div>
    </div>
  )
}
