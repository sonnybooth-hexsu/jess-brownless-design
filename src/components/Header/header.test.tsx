/* eslint no-undef: 0 */

import React from "react"
import { render, cleanup } from "@testing-library/react"
import { queryAllByTestId } from "@testing-library/dom"
import { Header } from "./header"

const testNavLinks = [
  {
    name: `Link 1`,
    nameShort: `Link 1`,
    page: `link-1`,
    subLinks: [
      {
        name: `Sub Link 1`,
        page: `Sub link-1`,
        id: 1,
      },
    ],
    id: 1,
  },
  {
    name: `Link 2`,
    nameShort: `Link 2`,
    page: `link-2`,
    subLinks: [],
    id: 2,
  },
  {
    name: `Link 3`,
    nameShort: `Link 3`,
    page: `link-3`,
    subLinks: [],
    id: 3,
  },
]

const noneSubLinks = testNavLinks.filter(item => !(item.subLinks.length > 0))
const hasSubLinks = testNavLinks.filter(item => item.subLinks.length > 0)

afterEach(() => {
  cleanup()
})

describe("<Header />", () => {
  it("should match snapshot", () => {
    const { container } = render(
      <Header
        siteTitle="Test Site"
        navLinks={testNavLinks}
        navToggle={() => Function}
        productItems={{}}
        openMenu={0}
        menuToggle={() => Function}
        page="test"
      />
    )

    expect(container).toMatchSnapshot()
  })

  it("should have the length of links passed through", () => {
    const { container } = render(
      <Header
        siteTitle="Test Site"
        navLinks={testNavLinks}
        navToggle={() => Function}
        productItems={{}}
        openMenu={0}
        menuToggle={() => Function}
        page="test"
      />
    )

    expect(queryAllByTestId(container, "headerNavigationLink")).toHaveLength(
      noneSubLinks.length
    )
  })

  it("toggles subLink class", () => {
    const { container } = render(
      <Header
        siteTitle="Test Site"
        navLinks={testNavLinks}
        navToggle={() => Function}
        productItems={{}}
        openMenu={1}
        menuToggle={() => Function}
        page="test"
      />
    )

    const subLinks = queryAllByTestId(container, "subLinkMenu")
    expect(subLinks).toHaveLength(hasSubLinks.length)

    const sublinkMenuList = queryAllByTestId(subLinks[0], "subLinkMenuList")[0]

    expect(sublinkMenuList).toHaveClass("block")
  })
})
