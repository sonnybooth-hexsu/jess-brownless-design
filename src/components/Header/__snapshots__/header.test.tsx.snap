// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should match snapshot 1`] = `
<div>
  <div
    class="
      bg-Brand-green
      text-white
      
      
       headerContainer fadedBorder"
  >
    <div
      class="header container"
    >
      <a
        class="headerLogo"
        href="/"
      >
        <img
          alt="Test Site"
          class="headerLogoImage"
          src="test-file-stub"
        />
      </a>
      <svg
        class="headerMenuIcon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="3"
          x2="21"
          y1="12"
          y2="12"
        />
        <line
          x1="3"
          x2="21"
          y1="6"
          y2="6"
        />
        <line
          x1="3"
          x2="21"
          y1="18"
          y2="18"
        />
      </svg>
      <nav
        class="headerNavigation"
      >
        <ol
          class="headerNavigationList"
        >
          <li
            data-testid="subLinkMenu"
            tabindex="0"
          >
            <span
              class="headerNavigationSubMenu"
            >
              Link 1
              <svg
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="6 9 12 15 18 9"
                />
              </svg>
            </span>
            <div
              class="hidden"
              data-testid="subLinkMenuList"
            >
              <div
                class="container grid grid-cols-3 gap-6 mx-auto"
              >
                <a
                  href="Sub link-1"
                >
                  Sub Link 1
                </a>
              </div>
            </div>
          </li>
          <li
            class="headerNavigationLinkItem "
            data-testid="headerNavigationLink"
          >
            <a
              class="headerNavigationLink"
              href="link-2"
            >
              <span
                class="xl:hidden"
              >
                Link 2
              </span>
              <span
                class="hidden xl:block"
              >
                Link 2
              </span>
            </a>
          </li>
          <li
            class="headerNavigationLinkItem "
            data-testid="headerNavigationLink"
          >
            <a
              class="headerNavigationLink"
              href="link-3"
            >
              <span
                class="xl:hidden"
              >
                Link 3
              </span>
              <span
                class="hidden xl:block"
              >
                Link 3
              </span>
            </a>
          </li>
          <li
            class="headerNavigationLinkItem "
          >
            <a
              class="headerNavigationLink"
              href="/contact"
            >
              Contact
            </a>
          </li>
        </ol>
      </nav>
    </div>
  </div>
</div>
`;
